File: Task001.py
#Silalak Kaewjanpet 610510670
import math 
 
class Point(): 
    def __init__(self, x, y): 
        self.x = x 
        self.y = y 
  
def dist(p1, p2): 
    return math.sqrt((p1.x - p2.x) * 
                     (p1.x - p2.x) +
                     (p1.y - p2.y) * 
                     (p1.y - p2.y))  
  
def bruteForce(P, N): 
    min_val = float('inf')  
    for i in range(N): 
        for j in range(i + 1, N): 
            if dist(P[i], P[j]) < min_val: 
                min_val = dist(P[i], P[j]) 
  
    return min_val 
  
def stripClosest(strip, size, d): 
      
   
    min_val = d  
  
    strip.sort(key = lambda point: point.y)  
  
    
    for i in range(size): 
        j = i + 1
        while j < size and (strip[j].y - 
                            strip[i].y) < min_val: 
            min_val = dist(strip[i], strip[j]) 
            j += 1
  
    return min_val  
  

def closestUtil(P, N): 
    
    if N <= 3:  
        return bruteForce(P, N)  
    
    mid = N // 2
    midPoint = P[mid] 
  
    dl = closestUtil(P[:mid], mid) 
    dr = closestUtil(P[mid:], N - mid)  
  
    d = min(dl, dr) 
  
    strip = []  
    for i in range(N):

        if abs(P[i].x - midPoint.x) < d:  
            strip.append(P[i]) 

    return min(d, stripClosest(strip, len(strip), d)) 
  
def closest(P, N): 

    P.sort(key = lambda point: point.x)
  
    return closestUtil(P, N) 
  

P = []
N = int(input())
for i in range(N):
    x, y = map(int,input().split())
    P.append(Point(x,y)) 
if closest(P, N) < 10000:
    print("%.4f"%(closest(P, N)))
else:
    print("INFINITY")

 

File: Task001.cpp
//Silalak Kaewjanpet sec 001
//610510670

#include <iostream>
using namespace std;
#include <iomanip>
#include <cstdlib> 
#include <bits/stdc++.h> 
  
class Point
{ 
	public: 
	int x, y; 
}; 


int compareX(const void* a, const void* b) 
{ 
	Point *p1 = (Point *)a, *p2 = (Point *)b; 
	return (p1->x - p2->x); 
} 


int compareY(const void* a, const void* b) 
{ 
	Point *p1 = (Point *)a, *p2 = (Point *)b; 
	return (p1->y - p2->y); 
} 

float dist(Point p1, Point p2) 
{ 
	return sqrt( (p1.x - p2.x)*(p1.x - p2.x) + 
				(p1.y - p2.y)*(p1.y - p2.y) 
			); 
} 

float bruteForce(Point P[], int n) 
{ 
	float min = FLT_MAX; 
	for (int i = 0; i < n; ++i) 
		for (int j = i+1; j < n; ++j) 
			if (dist(P[i], P[j]) < min) 
				min = dist(P[i], P[j]); 
	return min; 
} 

float min(float x, float y) 
{ 
	return (x < y)? x : y; 
} 


float stripClosest(Point strip[], int size, float d) 
{ 
	float min = d; // Initialize the minimum distance as d 

	qsort(strip, size, sizeof(Point), compareY); 

	for (int i = 0; i < size; ++i) 
		for (int j = i+1; j < size && (strip[j].y - strip[i].y) < min; ++j) 
			if (dist(strip[i],strip[j]) < min) 
				min = dist(strip[i], strip[j]); 

	return min; 
} 

float closestUtil(Point P[], int n) 
{ 
	if (n <= 3) 
		return bruteForce(P, n); 

	// Find the middle point 
	int mid = n/2; 
	Point midPoint = P[mid]; 

	float dl = closestUtil(P, mid); 
	float dr = closestUtil(P + mid, n - mid); 

	// Find the smaller of two distances 
	float d = min(dl, dr); 

	Point strip[n]; 
	int j = 0; 
	for (int i = 0; i < n; i++)
        { 
		    if (abs(P[i].x - midPoint.x) < d) 
			    strip[j] = P[i], j++; 
        }
	return min(d, stripClosest(strip, j, d) ); 
} 

float closest(Point P[], int n) 
{ 
	qsort(P, n, sizeof(Point), compareX); 

	return closestUtil(P, n); 
} 

// Driver code 
int main() 
{ 
    int N;      //จำนวนจุด
    cin >> N;

	int P[N*2];
    for(int i=0; i<N; i++)
	{
		cin >> P[];

	}
	Point P[];
	//Point P[] = {{0, 2}, {6, 67}, {43, 71}, {39, 107}, {189, 140}}; 
	int n = N;
    cout << showpoint;
    cout << fixed; 
	cout << setprecision(4) << closest(P, n); 
	return 0; 
} 

// This is code is contributed by rathbhupendra 
